lib/tbot/bot/bot.go

```diff
@@ -0,0 +1,371 @@
+				return trace.Wrap(err, "service(%s)", svc.String())
+			}
+
+			log.InfoContext(groupCtx, "Service finished")
+			return nil
+		})
+	}
+	return group.Wait()
+}
+
+func (b *Bot) buildServices(ctx context.Context) ([]Service, func(), error) {
+	startedAt := time.Now().UTC()
+	services := make([]Service, 0, len(b.cfg.Services))
+
+	var closers []func()
+	closeFn := func() {
+		for _, fn := range closers {
+			fn()
+		}
+	}
+
+	// Build internal services and dependencies.
+	statusRegistry := readyz.NewRegistry()
+	reloadBroadcaster := b.buildReloadBroadcaster(ctx)
+
+	resolver, err := b.buildResolver(ctx)
+	if err != nil {
+		return nil, closeFn, trace.Wrap(err, "building resolver")
+	}
+
+	clientBuilder, err := b.buildClientBuilder(resolver)
+	if err != nil {
+		return nil, closeFn, trace.Wrap(err, "building client builder")
+	}
+
+	identityService, closeIdentityService, err := b.buildIdentityService(
+		ctx,
+		reloadBroadcaster,
+		clientBuilder,
+		statusRegistry,
+	)
+	if err != nil {
+		return nil, closeFn, trace.Wrap(err, "building identity service")
+	}
+	services = append(services, identityService)
+	closers = append(closers, closeIdentityService)
+
+	heartbeatService, err := b.buildHeartbeatService(
+		identityService,
+		startedAt,
+		statusRegistry,
+	)
+	if err != nil {
+		return nil, closeFn, trace.Wrap(err, "building heartbeat service")
+	}
+	services = append(services, heartbeatService)
+
+	caRotationService, err := b.buildCARotationService(
+		reloadBroadcaster,
+		identityService,
+		statusRegistry,
+	)
+	if err != nil {
+		return nil, closeFn, trace.Wrap(err, "building CA rotation service")
+	}
+	services = append(services, caRotationService)
+
+	proxyPinger, err := b.buildProxyPinger(identityService)
+	if err != nil {
+		return nil, closeFn, trace.Wrap(err, "building proxy pinger")
+	}
+
+	identityGenerator, err := identityService.GetGenerator()
+	if err != nil {
+		return nil, closeFn, trace.Wrap(err, "building identity generator")
+	}
+
+	// Build user services.
+	for idx, buildService := range b.cfg.Services {
+		reloadCh, unsubscribe := reloadBroadcaster.Subscribe()
+		closers = append(closers, unsubscribe)
+
+		service, err := buildService(ServiceDependencies{
+			Client:             identityService.GetClient(),
+			Resolver:           resolver,
+			Logger:             b.cfg.Logger,
+			ClientBuilder:      clientBuilder,
+			IdentityGenerator:  identityGenerator,
+			ProxyPinger:        proxyPinger,
+			BotIdentity:        identityService.GetIdentity,
+			BotIdentityReadyCh: identityService.Ready(),
+			ReloadCh:           reloadCh,
+			StatusRegistry:     statusRegistry,
+		})
+		if err != nil {
+			return nil, closeFn, trace.Wrap(err, "building service [%d]", idx)
+		}
+		services = append(services, service)
+	}
+	return services, closeFn, nil
+}
+
+func (b *Bot) buildReloadBroadcaster(ctx context.Context) *internal.ChannelBroadcaster {
+	broadcaster := internal.NewChannelBroadcaster()
```

Comment:
How is this going to work in the one-shot mode?

Reply:
I'm not sure if this is what you meant, but I've fixed the possible goroutine leak if the context passed to `OneShot` is never cancelled: https://github.com/gravitational/teleport/pull/56658/commits/706887f952b86e0112090c569a8e277e48103dfb

Reply:
I meant exactly this. I was like hey, this is gonna run forever. :) Thanks.